packages/mobile/src/store/offline-downloads/sagas/downloadQueueSagas/syncCollectionWorker.ts

```diff
@@ -0,0 +1,222 @@
+import type { DownloadReason, ID } from '@audius/common'
+import {
+  reachabilityActions,
+  cacheCollectionsSelectors,
+  accountSelectors,
+  getContext
+} from '@audius/common'
+import { difference } from 'lodash'
+import moment from 'moment'
+import { call, put, race, select, take } from 'typed-redux-saga'
+
+import { DOWNLOAD_REASON_FAVORITES } from 'app/services/offline-downloader'
+import { dispatch } from 'app/store/store'
+
+import {
+  getCollectionSyncStatus,
+  getOfflineCollectionMetadata,
+  getOfflineTrackMetadata
+} from '../../selectors'
+import type { CollectionId, OfflineItem } from '../../slice'
+import {
+  completeCollectionSync,
+  CollectionSyncStatus,
+  errorCollectionSync,
+  cancelCollectionSync,
+  requestDownloadQueuedItem,
+  redownloadOfflineItems,
+  addOfflineItems,
+  removeOfflineItems,
+  startCollectionSync
+} from '../../slice'
+
+const { SET_UNREACHABLE } = reachabilityActions
+const { getUserId } = accountSelectors
+const { getCollection } = cacheCollectionsSelectors
+
+const isTrackFavoriteReason = (downloadReason: DownloadReason) =>
+  downloadReason.is_from_favorites &&
+  downloadReason.collection_id === DOWNLOAD_REASON_FAVORITES
+
+function* shouldCancelSync(collectionId: CollectionId) {
+  while (true) {
+    yield* take(removeOfflineItems.type)
+    const syncStatus = yield* select(getCollectionSyncStatus, collectionId)
+    if (!syncStatus) return true
+  }
+}
+
+export function* syncCollectionWorker(collectionId: CollectionId) {
+  yield* put(startCollectionSync({ id: collectionId }))
+
+  const { syncCollection, cancel, unreachable } = yield* race({
+    syncCollection: call(syncCollectionAsync, collectionId),
+    cancel: call(shouldCancelSync, collectionId),
+    unreachable: take(SET_UNREACHABLE)
+  })
+
+  if (cancel) {
+    yield* put(requestDownloadQueuedItem())
+  } else if (unreachable) {
+    yield* put(cancelCollectionSync({ id: collectionId }))
+  } else if (syncCollection === CollectionSyncStatus.ERROR) {
+    yield* put(errorCollectionSync({ id: collectionId }))
+    yield* put(requestDownloadQueuedItem())
+  } else if (syncCollection === CollectionSyncStatus.SUCCESS) {
+    yield* put(completeCollectionSync({ id: collectionId }))
+    yield* put(requestDownloadQueuedItem())
+  }
+}
+
+function* syncCollectionAsync(collectionId: CollectionId) {
+  if (collectionId === DOWNLOAD_REASON_FAVORITES) {
+    return yield* call(syncFavoritesCollection)
+  } else {
+    return yield* call(syncCollection, collectionId)
+  }
+}
+
+function* syncFavoritesCollection() {
+  const currentUserId = yield* select(getUserId)
+  const apiClient = yield* getContext('apiClient')
+  if (!currentUserId) return CollectionSyncStatus.ERROR
+  const offlineTrackMetadata = yield* select(getOfflineTrackMetadata)
+
+  const latestFavoritedTracks = yield* call(
```

Comment:
It's important that we return null if the getFavorites call fails or gives no response and that we stop the favorites sync in that case. Otherwise if the API call fails, we'll get `latestFavoritedTracks = []` and trackIdsToRemove will be all downloaded tracks. The `[]` return value needs to be reserved for the case where we got a successful response but the user has no favorited tracks on discovery.

Reply:
I think you've handled this properly in syncCollection as we return null from the APIClient and also have checks for empty data.

Reply:
yes wow great find! updating